src/hotspot/share/interpreter/linkResolver.cpp

```diff
@@ -1355,6 +1355,15 @@ void LinkResolver::runtime_resolve_virtual_method(CallInfo& result,
   // a missing receiver might result in a bogus lookup.
   assert(resolved_method->method_holder()->is_linked(), "must be linked");
 
+  // Receriver should be compatible with resolved klass, i.e. it must be a type of resolved klass
+  // or a subtype of resolved klass(receiver instanceof resolved_klass)
+  if (!recv.is_null() && !recv()->is_a(resolved_klass)) {
```

Comment:
It might be marginally cheaper to use `recv_klass->is_subtype_of(resolved_klass)` given you already have it available.